java.lang.AssertionError: not reached at com.yandex.yatagan.core.model.impl.ConditionalHoldingModelImpl$ConditionalWithFlavorConstraintsModelImpl.toString(ConditionalHoldingModelImpl.kt:70)

```
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.jetbrains.kotlin.kapt3.base.AnnotationProcessingKt.doAnnotationProcessing(annotationProcessing.kt:90)
	at org.jetbrains.kotlin.kapt3.base.AnnotationProcessingKt.doAnnotationProcessing$default(annotationProcessing.kt:31)
	at org.jetbrains.kotlin.kapt3.base.Kapt.kapt(Kapt.kt:47)
	... 33 more
Caused by: com.sun.tools.javac.processing.AnnotationProcessingError: java.lang.AssertionError: not reached
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:997)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:901)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1227)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1340)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1258)
	... 39 more
Caused by: java.lang.AssertionError: not reached
	at com.yandex.yatagan.core.model.impl.ConditionalHoldingModelImpl$ConditionalWithFlavorConstraintsModelImpl.toString(ConditionalHoldingModelImpl.kt:70)
	at com.yandex.yatagan.core.model.impl.ConditionalHoldingModelImpl$ConditionalWithFlavorConstraintsModelImpl.toString(ConditionalHoldingModelImpl.kt:57)
	at com.yandex.yatagan.validation.format.Format__FormatKt.append(format.kt:130)
	at com.yandex.yatagan.validation.format.Format.append(Unknown Source)
	at com.yandex.yatagan.validation.format.Strings$Errors.variantMatchingAmbiguity-vkkwLqc(Strings.kt:296)
	at com.yandex.yatagan.core.graph.impl.ConditionScopeForKt.VariantMatch(conditionScopeFor.kt:75)
	at com.yandex.yatagan.core.graph.impl.bindings.InjectConstructorProvisionBindingImpl$variantMatch$2.invoke(InjectConstructorProvisionBindingImpl.kt:42)
	at com.yandex.yatagan.core.graph.impl.bindings.InjectConstructorProvisionBindingImpl$variantMatch$2.invoke(InjectConstructorProvisionBindingImpl.kt:42)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at com.yandex.yatagan.core.graph.impl.bindings.InjectConstructorProvisionBindingImpl.getVariantMatch(InjectConstructorProvisionBindingImpl.kt:42)
	at com.yandex.yatagan.core.graph.impl.bindings.ConditionalBindingMixin.getConditionScope-s2kRxq8(mixins.kt:126)
	at com.yandex.yatagan.core.graph.impl.bindings.InjectConstructorProvisionBindingImpl$dependencies$2.invoke(InjectConstructorProvisionBindingImpl.kt:48)
	at com.yandex.yatagan.core.graph.impl.bindings.InjectConstructorProvisionBindingImpl$dependencies$2.invoke(InjectConstructorProvisionBindingImpl.kt:47)
	at kotlin.SafePublicationLazyImpl.getValue(LazyJVM.kt:107)
	at com.yandex.yatagan.core.graph.impl.bindings.InjectConstructorProvisionBindingImpl.getDependencies(InjectConstructorProvisionBindingImpl.kt:47)
	at com.yandex.yatagan.core.graph.impl.BindingGraphImpl.<init>(BindingGraphImpl.kt:202)
	at com.yandex.yatagan.core.graph.impl.BindingGraphImpl.<init>(BindingGraphImpl.kt)
	at com.yandex.yatagan.core.graph.impl.BindingGraphImpl$3.invoke(BindingGraphImpl.kt:155)
	at com.yandex.yatagan.core.graph.impl.BindingGraphImpl$3.invoke(BindingGraphImpl.kt:154)
	at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:210)
	at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:787)
	at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:817)
	at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:808)
	at com.yandex.yatagan.core.graph.impl.BindingGraphImpl.<init>(BindingGraphImpl.kt:165)
	at com.yandex.yatagan.core.graph.impl.BindingGraphImpl.<init>(BindingGraphImpl.kt)
	at com.yandex.yatagan.core.graph.impl.BindingGraphImpl.<init>(BindingGraphImpl.kt:51)
	at com.yandex.yatagan.core.graph.impl.BindingGraphKt.BindingGraph(BindingGraph.kt:30)
	at com.yandex.yatagan.processor.common.ProcessKt.process(process.kt:52)
	at com.yandex.yatagan.processor.jap.JapYataganProcessor.process(JapYataganProcessor.kt:80)
	at org.jetbrains.kotlin.kapt3.base.incremental.IncrementalProcessor.process(incrementalProcessors.kt:90)
	at org.jetbrains.kotlin.kapt3.base.ProcessorWrapper.process(annotationProcessing.kt:197)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:985)
	... 43 more
```

--- a/core/model/impl/src/main/kotlin/VariantImpl.kt
+++ b/core/model/impl/src/main/kotlin/VariantImpl.kt
@@ -75,8 +75,8 @@ internal class VariantImpl private constructor(
         representation = {
             when {
                 childContext != null -> {
-                    val (dimension, _) = parts.entries.find { (_, flavor) ->
-                        childContext == flavor
+                    val (dimension, _) = parts.entries.find { (_, flavors) ->
+                        childContext in flavors
                     }!!
                     append("{.., ")
                     appendChildContextReference(reference = buildRichString {

--- a/testing/tests/src/test/kotlin/ConditionsTest.kt
+++ b/testing/tests/src/test/kotlin/ConditionsTest.kt
@@ -333,6 +333,29 @@ class ConditionsTest(
     }
 
     @Test
+    fun `invalid variant filter declaration`() {
+        includeFromSourceSet(flavors)
+        includeFromSourceSet(features)
+
+        givenKotlinSource("test.TestCase", """
+            import com.yandex.yatagan.*
+            import javax.inject.*
+
+            @Conditionals(
+                Conditional(onlyIn = [ProductType.Browser::class], value = [Conditions.FeatureA::class]),
+                Conditional(onlyIn = [ProductType.Browser::class, ProductType.SearchApp::class], value = [Conditions.FeatureA::class]),
+            )
+            class TestClass @Inject constructor()
+
+            @Component(variant = [ProductType.Browser::class]) interface TestComponent {
+                val tc: Optional<TestClass>
+            }
+        """.trimIndent())
+
+        compileRunAndValidate()
+    }
+
+    @Test
     fun `@Binds with multiple alternatives`() {
         includeFromSourceSet(features)
         givenKotlinSource("test.TestCase", """

--- a/testing/tests/src/test/resources/golden/ConditionsTest/invalid-variant-filter-declaration.golden.txt
+++ b/testing/tests/src/test/resources/golden/ConditionsTest/invalid-variant-filter-declaration.golden.txt
@@ -1,0 +1,10 @@
+error: Variant matching ambiguity. Variant filters (onlyIn)
+    [component-flavor test.ProductType.Browser] and
+    [component-flavor test.ProductType.Browser, component-flavor test.ProductType.SearchApp]
+may both match at the same time.
+Encountered:
+  in graph for root-component test.TestComponent
+  in entry-point getTc: [Optional] test.TestClass
+                        ^-[*1]-------------------
+  here: [1*] inject-constructor test.TestClass()
+             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

--- a/validation/format/src/main/kotlin/Strings.kt
+++ b/validation/format/src/main/kotlin/Strings.kt
@@ -288,13 +288,16 @@ object Strings {
             append("No flavor is declared for `").append(dimension).append("` in a variant")
         }.toError()
 
+        @Covered
         fun variantMatchingAmbiguity(
             one: ConditionalHoldingModel.ConditionalWithFlavorConstraintsModel,
             two: ConditionalHoldingModel.ConditionalWithFlavorConstraintsModel,
         ) = buildRichString {
             color = TextColor.Inherit
-            append("Variant matching ambiguity: `").append(one).append("` vs `")
-                .append(two).append("` could not be resolved")
+            appendLine("Variant matching ambiguity. Variant filters (onlyIn)")
+            append(Indent).append(one.onlyIn).appendLine(" and")
+            append(Indent).append(two.onlyIn).appendLine()
+            append("may both match at the same time.")
         }.toError()
 
 
@@ -690,6 +693,7 @@ object Strings {
             append("Duplicate declared as `").append(factory).append('`')
         }.toNote()
 
+        @Covered
         fun conflictingFactory(factoryMethod: Method) = buildRichString {
             color = TextColor.Inherit
             append("Duplicate factory: `").append(factoryMethod).append('`')

--- a/validation/format/src/main/kotlin/format.kt
+++ b/validation/format/src/main/kotlin/format.kt
@@ -14,7 +14,8 @@
  * limitations under the License.
  */
 
-@file:[JvmMultifileClass JvmName("Format") ]
+@file:[JvmMultifileClass JvmName("Format")]
+
 package com.yandex.yatagan.validation.format
 
 import com.yandex.yatagan.core.graph.bindings.AliasBinding
@@ -105,7 +106,7 @@ inline fun BaseBinding.bindingModelRepresentation(
 }
 
 fun BaseBinding.findChildContextNode(childContext: MayBeInvalid?): NodeDependency? {
-    return when(childContext) {
+    return when (childContext) {
         is BaseBinding -> accept(object : BaseBinding.Visitor<NodeDependency?> {
             override fun visitAlias(alias: AliasBinding) = alias.source.takeIf { it == childContext.target }
             override fun visitBinding(binding: Binding) = binding.dependencies.find { it.node == childContext.target }
@@ -126,13 +127,23 @@ fun RichStringBuilder.appendChildContextReference(
 }
 
 fun RichStringBuilder.append(any: Any?): RichStringBuilder = apply {
-    when(any) {
+    when (any) {
         is MayBeInvalid -> append(any.toString(childContext = null))
         is CharSequence -> append(any)
         is ConditionExpression<*> -> append(any.toString(childContext = null))
         null -> append(Unresolved)
         else -> append(any.toString())
     }
+}
+
+fun RichStringBuilder.append(collection: Collection<Any?>): RichStringBuilder = apply {
+    append('[')
+    val lastIndex = collection.size - 1
+    collection.forEachIndexed { index, any ->
+        append(any)
+        if (index != lastIndex) append(", ")
+    }
+    append(']')
 }
 
 fun ConditionExpression<*>.toString(childContext: MayBeInvalid?): RichString = buildRichString {

